### ISSUE TITLE
コース1を完走する

### ISSUE THREAD

**Issue by nonanonno** [nonanonno/CarND-Behavioral-Cloning-P3#1]

**Comment by nonanonno** [comment]
# 試したこと
## データセット
### 画像
- 反時計回りに一周
- できるだけ真ん中を走るように
- 中心と左右のカメラ画像
- 320x160
- RGB
### annotation
- steering

## preprocess
- 中心のカメラに対する`steering`はそのまま、左であれば`steering + 0.2`、右であれば`steering - 0.2`とした
- データ拡張として左右反転を追加。steeringも反転
- 合計
    - train 4694
    - valid 1174

## ネットワーク
- Keras
- [nvidiaのやつ](https://devblogs.nvidia.com/deep-learning-self-driving-cars/)を少し修正して使用
- Conv2Dはエラったので1つ省略
- Croppingや正規化もKerasでやっているのでここで記述


|layer|description|output|
|:-:|:-:|:-:|
|Cropping2D|上50pix,下20pixを削除|90x320x3|
|Lambda|-0.5 ~ 0.5の値になるように正規化|90x320x3|
|Conv2D|5x5 kernel, 24 chns, relu, valid|86x316x24|
|MaxPooling2D|stride (2,2)|43x158x24|
|Conv2D|5x5 kernel, 36 chns, relu, valid|39x154x36|
|MaxPooling2D|stride (2,2)|19x77x36|
|Conv2D|5x5 kernel, 48 chns, relu, valid|15x73x48|
|MaxPooling2D|stride (2,2)|7x36x48|
|Conv2D|3x3 kernel, 64 chns, relu, valid|5x34x64|
|MaxPooling2D|stride (2,2)|2x17x64|
|Flatten||2176|
|Dence|100 nerrons|100|
|Dence|50 nerrons|50|
|Dence|10 nerrons|10|
|Dence|output|1|


## 学習
- validation_split: 0.2
- shuffle: true
- epochs: 20
- loss: mse
- optimizer: adam

## 結果
時計回り、反時計回り共に完走

**Comment by nonanonno** [comment]
## ToDo
- 学習曲線のプロット
- 走行結果の録画

**Comment by nonanonno** [comment]
* trainingデータを各epoch前にshuffle
    * overfittingを抑制

![track1](https://user-images.githubusercontent.com/10045065/46645325-60eae900-cbbf-11e8-9f5f-d026bdec5dea.png)

* ログ
```
Train on 4694 samples, validate on 1174 samples
Epoch 1/1
4694/4694 [==============================] - 5s - loss: 0.0222 - val_loss: 0.0125
Train on 4694 samples, validate on 1174 samples
Epoch 2/2
4694/4694 [==============================] - 5s - loss: 0.0098 - val_loss: 0.0084
Train on 4694 samples, validate on 1174 samples
Epoch 3/3
4694/4694 [==============================] - 5s - loss: 0.0062 - val_loss: 0.0051
Train on 4694 samples, validate on 1174 samples
Epoch 4/4
4694/4694 [==============================] - 5s - loss: 0.0044 - val_loss: 0.0038
Train on 4694 samples, validate on 1174 samples
Epoch 5/5
4694/4694 [==============================] - 5s - loss: 0.0041 - val_loss: 0.0040
Train on 4694 samples, validate on 1174 samples
Epoch 6/6
4694/4694 [==============================] - 5s - loss: 0.0032 - val_loss: 0.0028
Train on 4694 samples, validate on 1174 samples
Epoch 7/7
4694/4694 [==============================] - 5s - loss: 0.0028 - val_loss: 0.0025
Train on 4694 samples, validate on 1174 samples
Epoch 8/8
4694/4694 [==============================] - 5s - loss: 0.0024 - val_loss: 0.0025
Train on 4694 samples, validate on 1174 samples
Epoch 9/9
4694/4694 [==============================] - 5s - loss: 0.0021 - val_loss: 0.0024
Train on 4694 samples, validate on 1174 samples
Epoch 10/10
4694/4694 [==============================] - 5s - loss: 0.0020 - val_loss: 0.0033
Train on 4694 samples, validate on 1174 samples
Epoch 11/11
4694/4694 [==============================] - 5s - loss: 0.0018 - val_loss: 0.0020
Train on 4694 samples, validate on 1174 samples
Epoch 12/12
4694/4694 [==============================] - 5s - loss: 0.0018 - val_loss: 0.0020
Train on 4694 samples, validate on 1174 samples
Epoch 13/13
4694/4694 [==============================] - 5s - loss: 0.0014 - val_loss: 0.0011
Train on 4694 samples, validate on 1174 samples
Epoch 14/14
4694/4694 [==============================] - 5s - loss: 0.0012 - val_loss: 0.0011
Train on 4694 samples, validate on 1174 samples
Epoch 15/15
4694/4694 [==============================] - 5s - loss: 0.0012 - val_loss: 0.0013
Train on 4694 samples, validate on 1174 samples
Epoch 16/16
4694/4694 [==============================] - 5s - loss: 0.0013 - val_loss: 0.0011
Train on 4694 samples, validate on 1174 samples
Epoch 17/17
4694/4694 [==============================] - 5s - loss: 0.0012 - val_loss: 0.0012
Train on 4694 samples, validate on 1174 samples
Epoch 18/18
4694/4694 [==============================] - 5s - loss: 0.0011 - val_loss: 0.0010
Train on 4694 samples, validate on 1174 samples
Epoch 19/19
4694/4694 [==============================] - 5s - loss: 0.0011 - val_loss: 0.0010
Train on 4694 samples, validate on 1174 samples
Epoch 20/20
4694/4694 [==============================] - 5s - loss: 0.0010 - val_loss: 8.4717e-04
```
